### PR TITLE
[FLINK-4035] Bump Kafka producer in Kafka sink to Kafka 0.10.0.0

### PR DESCRIPTION
**Pull request by radekg**
Hi everyone,

At The Weather Company we bumped into a problem while trying to use Flink with Kafka 0.10.x. This PR introduces the support for `FlinkKafkaConsumer010` and `FlinkKafkaProducer010`. Unit test coverage is provided and `mvn clean verify` passes.

The output is below:

```
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO]
[INFO] force-shading ...................................... SUCCESS [  1.311 s]
[INFO] flink .............................................. SUCCESS [  2.939 s]
[INFO] flink-annotations .................................. SUCCESS [  1.476 s]
[INFO] flink-shaded-hadoop ................................ SUCCESS [  0.152 s]
[INFO] flink-shaded-hadoop2 ............................... SUCCESS [  7.065 s]
[INFO] flink-shaded-include-yarn-tests .................... SUCCESS [  8.543 s]
[INFO] flink-shaded-curator ............................... SUCCESS [  0.112 s]
[INFO] flink-shaded-curator-recipes ....................... SUCCESS [  1.080 s]
[INFO] flink-shaded-curator-test .......................... SUCCESS [  0.210 s]
[INFO] flink-test-utils-parent ............................ SUCCESS [  0.126 s]
[INFO] flink-test-utils-junit ............................. SUCCESS [  2.019 s]
[INFO] flink-core ......................................... SUCCESS [ 34.501 s]
[INFO] flink-java ......................................... SUCCESS [ 26.266 s]
[INFO] flink-runtime ...................................... SUCCESS [04:57 min]
[INFO] flink-optimizer .................................... SUCCESS [  7.914 s]
[INFO] flink-clients ...................................... SUCCESS [  6.537 s]
[INFO] flink-streaming-java ............................... SUCCESS [ 37.732 s]
[INFO] flink-test-utils ................................... SUCCESS [  6.166 s]
[INFO] flink-scala ........................................ SUCCESS [ 24.626 s]
[INFO] flink-runtime-web .................................. SUCCESS [ 12.831 s]
[INFO] flink-examples ..................................... SUCCESS [  1.123 s]
[INFO] flink-examples-batch ............................... SUCCESS [ 11.919 s]
[INFO] flink-contrib ...................................... SUCCESS [  0.096 s]
[INFO] flink-statebackend-rocksdb ......................... SUCCESS [  7.770 s]
[INFO] flink-tests ........................................ SUCCESS [06:22 min]
[INFO] flink-streaming-scala .............................. SUCCESS [ 26.831 s]
[INFO] flink-streaming-connectors ......................... SUCCESS [  0.100 s]
[INFO] flink-connector-flume .............................. SUCCESS [  2.425 s]
[INFO] flink-libraries .................................... SUCCESS [  0.084 s]
[INFO] flink-table ........................................ SUCCESS [02:02 min]
[INFO] flink-connector-kafka-base ......................... SUCCESS [  4.604 s]
[INFO] flink-connector-kafka-0.8 .......................... SUCCESS [02:01 min]
[INFO] flink-connector-kafka-0.9 .......................... SUCCESS [02:38 min]
[INFO] flink-connector-kafka-0.10 ......................... SUCCESS [02:04 min]
[INFO] flink-connector-elasticsearch ...................... SUCCESS [ 19.310 s]
[INFO] flink-connector-elasticsearch2 ..................... SUCCESS [ 17.086 s]
[INFO] flink-connector-rabbitmq ........................... SUCCESS [  2.885 s]
[INFO] flink-connector-twitter ............................ SUCCESS [  2.649 s]
[INFO] flink-connector-nifi ............................... SUCCESS [  1.339 s]
[INFO] flink-connector-cassandra .......................... SUCCESS [01:21 min]
[INFO] flink-connector-redis .............................. SUCCESS [  5.738 s]
[INFO] flink-connector-filesystem ......................... SUCCESS [ 24.871 s]
[INFO] flink-batch-connectors ............................. SUCCESS [  0.103 s]
[INFO] flink-avro ......................................... SUCCESS [  9.788 s]
[INFO] flink-jdbc ......................................... SUCCESS [  4.839 s]
[INFO] flink-hadoop-compatibility ......................... SUCCESS [ 10.026 s]
[INFO] flink-hbase ........................................ SUCCESS [  2.938 s]
[INFO] flink-hcatalog ..................................... SUCCESS [  5.383 s]
[INFO] flink-examples-streaming ........................... SUCCESS [ 24.339 s]
[INFO] flink-gelly ........................................ SUCCESS [ 40.416 s]
[INFO] flink-gelly-scala .................................. SUCCESS [ 29.050 s]
[INFO] flink-gelly-examples ............................... SUCCESS [ 21.418 s]
[INFO] flink-python ....................................... SUCCESS [ 58.016 s]
[INFO] flink-ml ........................................... SUCCESS [ 57.270 s]
[INFO] flink-cep .......................................... SUCCESS [  6.984 s]
[INFO] flink-cep-scala .................................... SUCCESS [  8.521 s]
[INFO] flink-scala-shell .................................. SUCCESS [03:28 min]
[INFO] flink-quickstart ................................... SUCCESS [  1.248 s]
[INFO] flink-quickstart-java .............................. SUCCESS [  0.610 s]
[INFO] flink-quickstart-scala ............................. SUCCESS [  0.237 s]
[INFO] flink-storm ........................................ SUCCESS [ 14.975 s]
[INFO] flink-storm-examples ............................... SUCCESS [ 37.513 s]
[INFO] flink-streaming-contrib ............................ SUCCESS [  8.452 s]
[INFO] flink-tweet-inputformat ............................ SUCCESS [  3.075 s]
[INFO] flink-operator-stats ............................... SUCCESS [  6.521 s]
[INFO] flink-connector-wikiedits .......................... SUCCESS [ 18.022 s]
[INFO] flink-yarn ......................................... SUCCESS [  7.539 s]
[INFO] flink-dist ......................................... SUCCESS [ 11.453 s]
[INFO] flink-metrics ...................................... SUCCESS [  0.101 s]
[INFO] flink-metrics-dropwizard ........................... SUCCESS [  2.699 s]
[INFO] flink-metrics-ganglia .............................. SUCCESS [  1.320 s]
[INFO] flink-metrics-graphite ............................. SUCCESS [  1.188 s]
[INFO] flink-metrics-statsd ............................... SUCCESS [  2.271 s]
[INFO] flink-fs-tests ..................................... SUCCESS [ 27.916 s]
[INFO] flink-java8 ........................................ SUCCESS [ 12.209 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 37:24 min
[INFO] Finished at: 2016-07-12T15:18:39-04:00
[INFO] Final Memory: 234M/1833M
[INFO] ------------------------------------------------------------------------
```

The only thing not provided right in this moment in time, is the documentation updates. Not sure how to take on that one, some guidance would be appreciated.

What would be the best way to proceed with the contribution?

### CODE REVIEW

**Comment by tzulitai** [comment]
Hi @radekg , thank you for opening a PR for this!
From a first look it seems that there isn't much changes to the code of `flink-connector-kafka-0.9` and this PR. Also, from the original discussion / comments in the JIRA, the Kafka API doesn't seem to have changed between 0.9 and 0.10, so it might be possible to let the Kafka 0.9 connector use the 0.10 client by putting the Kafka 0.10 dependency into the user pom.

May I ask whether you have tried this approach out already? Also,
> At The Weather Company we bumped into a problem while trying to use Flink with Kafka 0.10.x.
What was the problem? If you can describe, it'll be helpful for deciding how we can proceed with this :) There's another contributor who was trying this out, I'll also try to ask for his feedback on this in the JIRA.

**Comment by radekg** [comment]
Sure, the problems are the following:

- https://github.com/apache/flink/pull/2231/commits/06936d7c5acc0897348019161c9ced4596a0a4dd#diff-aba21cf86694f3f2cd85e2e5e9b04972R305 in 0.9, `consumer.assign` (https://github.com/apache/flink/pull/2231/commits/06936d7c5acc0897348019161c9ced4596a0a4dd#diff-aba21cf86694f3f2cd85e2e5e9b04972R180) takes a `List`, in 0.10 it takes `Collection`
- for unit tests: https://github.com/apache/flink/pull/2231/commits/06936d7c5acc0897348019161c9ced4596a0a4dd#diff-ab65f3156ed8820677f3420152b78908R130, if we use 0.9 kafka version with 0.10 client, the concrete client tests fail as they catch wrong exception type in: https://github.com/TheWeatherCompany/flink/blob/06936d7c5acc0897348019161c9ced4596a0a4dd/flink-streaming-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java#L185

Silly stuff. Everything else works just fine. Fell free to reuse this stuff.

FYI: I'd be confused it I was to use a class indicating 0.9 when working with 0.10, that's the reason I assembled separate module. 0.9 is done and there's no future work required, it makes sense to have 0.10. Just my opinion.

**Comment by tzulitai** [comment]
Thank you for the description, @radekg .

I think the problems you mentioned should be solvable by working on the 0.9 connector to be just a bit more general, then users can simply manually use the 0.10 jars. However, you also have a point on the possible confusion. IMHO, I think it is redundant to have two connector modules with almost the same code, and it doesn't also seem feasible for code maintainability to keep adding modules for new Kafka versions even if they don't have changes in the API.

I think we'll need to loop in @rmetzger and @aljoscha to decide on how we can proceed with this. The solutions I currently see is to work on the 0.9 connector on the above problems so it can be compatible with the 0.10 API, and either rename the module to be `flink-connecto-kafka-0.10` (doesn't seem good because it'll be breaking user's pom's), or add information to the documentation on how to work with Kafka 0.10. Either way, in the long-run, we'll probably still need to sort out a way to better manage the connector codes in situations of new external system versions like this.

**Comment by aljoscha** [comment]
I agree with @tzulitai that it would be nice if you could have minimum code duplication in the long-run but it might not be possible with the current design of the consumers.

What about the new timestamps that were introduced in Kafka 0.10? This is also something that wouldn't work with the 0.9 consumer and could only be implemented for the 0.10-specific consumer, correct?

**Comment by tzulitai** [comment]
@aljoscha Ah right, for the new timestamps we will definitely need a 0.10-specific consumer. I think it makes sense to include a new module then.

This current PR does not add the new 0.10 timestamps ( @radekg correct me if I'm wrong here ), but I think we can add this as a separate follow up JIRA / PR afterwards, because it'll probably require changing some code in the `flink-connector-kafka-base` and the user-facing deserialization schemas that will need more discussion.

I'll find time this week to give this PR a test.

**Comment by radekg** [comment]
@tzulitai yes, this pr does not deal with 0.10 specific timestamps. It makes a simple consumer application work.

**Comment by tzulitai** [comment]
Hi @radekg , 
There was a recent change to the connector code to how the Kafka metrics are reported, so right now the PR has conflicts and can't be built. Would you like to rebase this PR on the current master branch so we can start testing it?

**Comment by radekg** [comment]
Merged with `upstream/master` and I'm getting this when running `mvn clean verify`:

```
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR :
[INFO] -------------------------------------------------------------
[ERROR] /Users/rad/dev/twc/flink/flink-streaming-connectors/flink-connector-kafka-0.10/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka010Fetcher.java:[30,69] cannot find symbol
  symbol:   class DefaultKafkaMetricAccumulator
  location: package org.apache.flink.streaming.connectors.kafka.internals.metrics
[ERROR] /Users/rad/dev/twc/flink/flink-streaming-connectors/flink-connector-kafka-0.10/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka010Fetcher.java:[105,17] constructor AbstractFetcher in class org.apache.flink.streaming.connectors.kafka.internals.AbstractFetcher<T,KPH> cannot be applied to given types;
  required: org.apache.flink.streaming.api.functions.source.SourceFunction.SourceContext<T>,java.util.List<org.apache.flink.streaming.connectors.kafka.internals.KafkaTopicPartition>,org.apache.flink.util.SerializedValue<org.apache.flink.streaming.api.functions.AssignerWithPeriodicWatermarks<T>>,org.apache.flink.util.SerializedValue<org.apache.flink.streaming.api.functions.AssignerWithPunctuatedWatermarks<T>>,org.apache.flink.streaming.api.operators.StreamingRuntimeContext,boolean
  found: org.apache.flink.streaming.api.functions.source.SourceFunction.SourceContext<T>,java.util.List<org.apache.flink.streaming.connectors.kafka.internals.KafkaTopicPartition>,org.apache.flink.util.SerializedValue<org.apache.flink.streaming.api.functions.AssignerWithPeriodicWatermarks<T>>,org.apache.flink.util.SerializedValue<org.apache.flink.streaming.api.functions.AssignerWithPunctuatedWatermarks<T>>,org.apache.flink.streaming.api.operators.StreamingRuntimeContext
  reason: actual and formal argument lists differ in length
[ERROR] /Users/rad/dev/twc/flink/flink-streaming-connectors/flink-connector-kafka-0.10/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka010Fetcher.java:[192,49] cannot find symbol
  symbol:   class DefaultKafkaMetricAccumulator
  location: class org.apache.flink.streaming.connectors.kafka.internal.Kafka010Fetcher<T>
[ERROR] /Users/rad/dev/twc/flink/flink-streaming-connectors/flink-connector-kafka-0.10/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka010Fetcher.java:[193,65] cannot find symbol
  symbol:   variable DefaultKafkaMetricAccumulator
  location: class org.apache.flink.streaming.connectors.kafka.internal.Kafka010Fetcher<T>
[INFO] 4 errors
[INFO] -------------------------------------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO]
[INFO] force-shading ...................................... SUCCESS [  1.210 s]
[INFO] flink .............................................. SUCCESS [  4.416 s]
[INFO] flink-annotations .................................. SUCCESS [  1.551 s]
[INFO] flink-shaded-hadoop ................................ SUCCESS [  0.162 s]
[INFO] flink-shaded-hadoop2 ............................... SUCCESS [  6.451 s]
[INFO] flink-shaded-include-yarn-tests .................... SUCCESS [  7.929 s]
[INFO] flink-shaded-curator ............................... SUCCESS [  0.110 s]
[INFO] flink-shaded-curator-recipes ....................... SUCCESS [  0.986 s]
[INFO] flink-shaded-curator-test .......................... SUCCESS [  0.200 s]
[INFO] flink-test-utils-parent ............................ SUCCESS [  0.111 s]
[INFO] flink-test-utils-junit ............................. SUCCESS [  2.417 s]
[INFO] flink-core ......................................... SUCCESS [ 37.825 s]
[INFO] flink-java ......................................... SUCCESS [ 23.620 s]
[INFO] flink-runtime ...................................... SUCCESS [06:25 min]
[INFO] flink-optimizer .................................... SUCCESS [ 12.698 s]
[INFO] flink-clients ...................................... SUCCESS [  9.795 s]
[INFO] flink-streaming-java ............................... SUCCESS [ 43.709 s]
[INFO] flink-test-utils ................................... SUCCESS [  9.363 s]
[INFO] flink-scala ........................................ SUCCESS [ 37.639 s]
[INFO] flink-runtime-web .................................. SUCCESS [ 19.749 s]
[INFO] flink-examples ..................................... SUCCESS [  1.006 s]
[INFO] flink-examples-batch ............................... SUCCESS [ 14.276 s]
[INFO] flink-contrib ...................................... SUCCESS [  0.104 s]
[INFO] flink-statebackend-rocksdb ......................... SUCCESS [ 10.938 s]
[INFO] flink-tests ........................................ SUCCESS [07:34 min]
[INFO] flink-streaming-scala .............................. SUCCESS [ 33.365 s]
[INFO] flink-streaming-connectors ......................... SUCCESS [  0.106 s]
[INFO] flink-connector-flume .............................. SUCCESS [  5.626 s]
[INFO] flink-libraries .................................... SUCCESS [  0.100 s]
[INFO] flink-table ........................................ SUCCESS [02:31 min]
[INFO] flink-connector-kafka-base ......................... SUCCESS [ 10.033 s]
[INFO] flink-connector-kafka-0.8 .......................... SUCCESS [02:06 min]
[INFO] flink-connector-kafka-0.9 .......................... SUCCESS [02:12 min]
[INFO] flink-connector-kafka-0.10 ......................... FAILURE [  0.197 s]
[INFO] flink-connector-elasticsearch ...................... SKIPPED
[INFO] flink-connector-elasticsearch2 ..................... SKIPPED
[INFO] flink-connector-rabbitmq ........................... SKIPPED
[INFO] flink-connector-twitter ............................ SKIPPED
[INFO] flink-connector-nifi ............................... SKIPPED
[INFO] flink-connector-cassandra .......................... SKIPPED
[INFO] flink-connector-redis .............................. SKIPPED
[INFO] flink-connector-filesystem ......................... SKIPPED
[INFO] flink-batch-connectors ............................. SKIPPED
[INFO] flink-avro ......................................... SKIPPED
[INFO] flink-jdbc ......................................... SKIPPED
[INFO] flink-hadoop-compatibility ......................... SKIPPED
[INFO] flink-hbase ........................................ SKIPPED
[INFO] flink-hcatalog ..................................... SKIPPED
[INFO] flink-examples-streaming ........................... SKIPPED
[INFO] flink-gelly ........................................ SKIPPED
[INFO] flink-gelly-scala .................................. SKIPPED
[INFO] flink-gelly-examples ............................... SKIPPED
[INFO] flink-python ....................................... SKIPPED
[INFO] flink-ml ........................................... SKIPPED
[INFO] flink-cep .......................................... SKIPPED
[INFO] flink-cep-scala .................................... SKIPPED
[INFO] flink-scala-shell .................................. SKIPPED
[INFO] flink-quickstart ................................... SKIPPED
[INFO] flink-quickstart-java .............................. SKIPPED
[INFO] flink-quickstart-scala ............................. SKIPPED
[INFO] flink-storm ........................................ SKIPPED
[INFO] flink-storm-examples ............................... SKIPPED
[INFO] flink-streaming-contrib ............................ SKIPPED
[INFO] flink-tweet-inputformat ............................ SKIPPED
[INFO] flink-operator-stats ............................... SKIPPED
[INFO] flink-connector-wikiedits .......................... SKIPPED
[INFO] flink-yarn ......................................... SKIPPED
[INFO] flink-dist ......................................... SKIPPED
[INFO] flink-metrics ...................................... SKIPPED
[INFO] flink-metrics-dropwizard ........................... SKIPPED
[INFO] flink-metrics-ganglia .............................. SKIPPED
[INFO] flink-metrics-graphite ............................. SKIPPED
[INFO] flink-metrics-statsd ............................... SKIPPED
[INFO] flink-fs-tests ..................................... SKIPPED
[INFO] flink-java8 ........................................ SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 25:46 min
[INFO] Finished at: 2016-07-22T21:52:55+02:00
[INFO] Final Memory: 159M/1763M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.1:compile (default-compile) on project flink-connector-kafka-0.10_2.10: Compilation failure: Compilation failure:
[ERROR] /Users/rad/dev/twc/flink/flink-streaming-connectors/flink-connector-kafka-0.10/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka010Fetcher.java:[30,69] cannot find symbol
[ERROR] symbol:   class DefaultKafkaMetricAccumulator
[ERROR] location: package org.apache.flink.streaming.connectors.kafka.internals.metrics
[ERROR] /Users/rad/dev/twc/flink/flink-streaming-connectors/flink-connector-kafka-0.10/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka010Fetcher.java:[105,17] constructor AbstractFetcher in class org.apache.flink.streaming.connectors.kafka.internals.AbstractFetcher<T,KPH> cannot be applied to given types;
[ERROR] required: org.apache.flink.streaming.api.functions.source.SourceFunction.SourceContext<T>,java.util.List<org.apache.flink.streaming.connectors.kafka.internals.KafkaTopicPartition>,org.apache.flink.util.SerializedValue<org.apache.flink.streaming.api.functions.AssignerWithPeriodicWatermarks<T>>,org.apache.flink.util.SerializedValue<org.apache.flink.streaming.api.functions.AssignerWithPunctuatedWatermarks<T>>,org.apache.flink.streaming.api.operators.StreamingRuntimeContext,boolean
[ERROR] found: org.apache.flink.streaming.api.functions.source.SourceFunction.SourceContext<T>,java.util.List<org.apache.flink.streaming.connectors.kafka.internals.KafkaTopicPartition>,org.apache.flink.util.SerializedValue<org.apache.flink.streaming.api.functions.AssignerWithPeriodicWatermarks<T>>,org.apache.flink.util.SerializedValue<org.apache.flink.streaming.api.functions.AssignerWithPunctuatedWatermarks<T>>,org.apache.flink.streaming.api.operators.StreamingRuntimeContext
[ERROR] reason: actual and formal argument lists differ in length
[ERROR] /Users/rad/dev/twc/flink/flink-streaming-connectors/flink-connector-kafka-0.10/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka010Fetcher.java:[192,49] cannot find symbol
[ERROR] symbol:   class DefaultKafkaMetricAccumulator
[ERROR] location: class org.apache.flink.streaming.connectors.kafka.internal.Kafka010Fetcher<T>
[ERROR] /Users/rad/dev/twc/flink/flink-streaming-connectors/flink-connector-kafka-0.10/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka010Fetcher.java:[193,65] cannot find symbol
[ERROR] symbol:   variable DefaultKafkaMetricAccumulator
[ERROR] location: class org.apache.flink.streaming.connectors.kafka.internal.Kafka010Fetcher<T>
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR]
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :flink-connector-kafka-0.10_2.10
```

Any advice?

**Comment by tzulitai** [comment]
The errors are due to some of the changes to `AbstractFetcher` in https://github.com/apache/flink/commit/41f58182289226850b23c61a32f01223485d4775. Some of the Kafka 0.9 connector code that has changed accordingly, so you'll probably need to reflect those changes in the Kafka 0.10 code too.

**Comment by radekg** [comment]
Thanks, running `verify` again.

**Comment by radekg** [comment]
Travis is going to run.

**Comment by tzulitai** [comment]
@radekg Thank you for the quick fix. I hope to find time over the weekend to test + review this, if not than early next week :)

**Comment by radekg** [comment]
Tests are failing for random setups on travis. Seems to be something scala related.

**Comment by tzulitai** [comment]
Sorry for the delay on the review. Reviewing + testing now ...

@radekg Yup the failure is unrelated to the changes here. Some of the tests are currently a bit flaky.

File: flink-streaming-connectors/flink-connector-kafka-0.10/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaTestEnvironmentImpl.java [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+	}
+
+	@Override
+	public void createTestTopic(String topic, int numberOfPartitions, int replicationFactor, Properties topicConfig) {
+		// create topic with one client
+		LOG.info("Creating topic {}", topic);
+
+		ZkUtils zkUtils = getZkUtils();
+		try {
+			AdminUtils.createTopic(zkUtils, topic, numberOfPartitions, replicationFactor, topicConfig, new kafka.admin.RackAwareMode.Enforced$());
```
Comment (tzulitai) [inline flink-streaming-connectors/flink-connector-kafka-0.10/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaTestEnvironmentImpl.java]: The proper usage of `RackAwareMode` here seems to be `kafka.admin.RackAwareMode.Enforced$.MODULE$` (this is how tests in Kafka use this). IntelliJ complains that `new kafka.admin.RackAwareMode.Enforced$()` has private access, I'm not sure why the build is passing on this though ...

File: flink-streaming-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java [hunk 2]
```diff
@@ -134,6 +134,10 @@
 	@Rule
 	public RetryRule retryRule = new RetryRule();
 
+	public String getExpectedKafkaVersion() {
+		return "0.9";
+	}
```
Comment (tzulitai) [inline flink-streaming-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java]: We should make this an abstract method in the abstract test base.

File: flink-streaming-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java [hunk 3]
```diff
@@ -186,7 +190,7 @@ public void runFailOnNoBrokerTest() throws Exception {
 			stream.print();
 			see.execute("No broker test");
 		} catch(RuntimeException re) {
-			if(kafkaServer.getVersion().equals("0.9")) {
+			if(kafkaServer.getVersion().equals(getExpectedKafkaVersion())) {
```
Comment (tzulitai) [inline flink-streaming-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java]: I think the original intent of this assert here was that 0.8 connector throws different exception messages than 0.9.
So adding the `getExpectedKafkaVersion()` is a bit confusing with respect to the test intent.
Perhaps we should remove the new `getExpectedKafkaVersion()` and simply change this to `kafkaServer.getVersion().equals("0.9") || kafkaServer.getVersion().equals("0.10")`?

File: flink-streaming-connectors/flink-connector-kafka-0.10/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer010.java [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+
+	private static final long serialVersionUID = 2324564345203409112L;
+
+	private static final Logger LOG = LoggerFactory.getLogger(FlinkKafkaConsumer010.class);
+
+	/**  Configuration key to change the polling timeout **/
+	public static final String KEY_POLL_TIMEOUT = "flink.poll-timeout";
+
+	/** Boolean configuration key to disable metrics tracking **/
+	public static final String KEY_DISABLE_METRICS = "flink.disable-metrics";
```
Comment (tzulitai) [inline flink-streaming-connectors/flink-connector-kafka-0.10/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer010.java]: This is redundant. It's already declared in the `FlinkKafkaConsumerBase`.

**Comment by tzulitai** [comment]
Finished first review of the code.

Let me summarize parts of the Kafka 0.10 API that requires us to have a separate module:

 1. `ConsumerRecord` in 0.10 has a new `ConsumerRecord#timestamp()` method to retrieve Kafka server-side timestamps. If we want to attach this timestamp to the records as the default event time in the future, we'd definitely need a separate module (0.10 timestamp feature not included in this PR).
 2. `PartitionMetaData` (used in `KafkaTestEnvironmentImpl`s) has a breaking change to the APIs for retrieving the info, so we can't simply bump the version either.

Other than the above, the rest of the code is the same between (or changes are irrelevant to Kafka API changes) the 0.9 connector.

**Comment by tzulitai** [comment]
@radekg One thing that's missing is update to the docs. You can find the Kafka connector documentation at `flink/docs/apis/streaming/connectors/kafka.md`. We'lll probably only need to update the Maven dependency table to include the 0.10 connector.

File: flink-streaming-connectors/flink-connector-kafka-0.10/pom.xml [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+		 xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+		 xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+
+	<modelVersion>4.0.0</modelVersion>
+
+	<parent>
+		<groupId>org.apache.flink</groupId>
+		<artifactId>flink-streaming-connectors</artifactId>
+		<version>1.1-SNAPSHOT</version>
```
Comment (tzulitai) [inline flink-streaming-connectors/flink-connector-kafka-0.10/pom.xml]: We've recently just bumped version to 1.2-SNAPSHOT.

**Comment by rmetzger** [comment]
Sorry for joining this discussion late. I've been on vacation.
I also stumbled across the code duplicates. I'll check out the code from this pull request and see if there's a good way of re-using most of the 0.9 connector code.

**Comment by rmetzger** [comment]
@radekg, are you okay with me using your pull request as a base for adding Kafka 0.10 to Flink?
I've started changing your code from the PR so that we don't need to copy so much code: https://github.com/rmetzger/flink/commits/kafka-0.10

**Comment by radekg** [comment]
@rmetzger it's absolutely fine to reuse the code. If I can help in any way, please let me know.

**Comment by rmetzger** [comment]
Okay, cool. Thank you. I'll probably open a pull request with your and my changes. I'll let you know so that you can help reviewing it.

**Comment by rmetzger** [comment]
@radekg I've opened a PR based on your code: https://github.com/apache/flink/issues/2369 feel free to review it.

**Comment by rmetzger** [comment]
@radekg I've added Kafka 0.10 support to Flink, that's why I closed this pull request. My change preserved your commit from this pull request. Thank you for the contribution!

**Pull request closed without merging**